### PR TITLE
[FLINK-12608][runtime] Add getVertexOrThrow and

### PR DESCRIPTION
**Pull request by eaglewatcherwb**
getResultPartitionOrThrow in SchedulingTopology to simplify code
structure

Change-Id: I33be923c3279755d360e95336199fb1bc3257f4f

## What is the purpose of the change

*Add getVertexOrThrow and getResultPartitionOrThrow in SchedulingTopology to simplify code structure.*

## Brief change log

  - *Add getVertexOrThrow and getResultPartitionOrThrow in SchedulingTopology.*

## Verifying this change

This change added tests and can be verified as follows:

  - Added unit tests in ExecutionGraphToSchedulingTopologyAdapterTest. 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/SchedulingTopology.java [hunk 1]
```diff
@@ -43,11 +43,27 @@
 	 */
 	Optional<SchedulingExecutionVertex> getVertex(ExecutionVertexID executionVertexId);
 
+	/**
+	 * Looks up the {@link SchedulingExecutionVertex} for the given {@link ExecutionVertexID}.
+	 *
+	 * @param executionVertexId identifying the respective scheduling vertex
+	 * @return The respective scheduling vertex or throw {@link Exception} if the vertex does not exist
+	 */
+	SchedulingExecutionVertex getVertexOrThrow(ExecutionVertexID executionVertexId);
```
Comment (GJL) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/SchedulingTopology.java]: Did you consider using a default method? https://docs.oracle.com/javase/tutorial/java/IandI/defaultmethods.html

```
default SchedulingExecutionVertex getVertexOrThrow(ExecutionVertexID executionVertexId) {
...
}

(This is not a review)
```

Comment (eaglewatcherwb) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/SchedulingTopology.java]: Emm, `default` is a good way to add new method in interfaces, since in some user cases it is not easy to add corresponding method in the existing implementations. But using `default` may have the risk of `existing implementations of an interface may compile without error or warning but fail at runtime` (as Item 21 in Effective Java), thus I prefer not to use `default` as far as possible.
Anyway, thanks a lot for providing this `default` proposal.

Comment (eaglewatcherwb) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/SchedulingTopology.java]: Emm, it seems there is no risk of current implementation. We could use `default`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/strategy/TestingSchedulingTopology.java [hunk 2]
```diff
@@ -210,8 +210,8 @@ private ProducerConsumerAllToAllConnectionBuilder(
 				for (TestingSchedulingExecutionVertex consumer : consumers) {
 					consumer.addConsumedPartition(resultPartition);
 					resultPartition.addConsumer(consumer);
-					resultPartitions.add(resultPartition);
 				}
+				resultPartitions.add(resultPartition);
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/strategy/TestingSchedulingTopology.java]: Change should be a separate commit

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/SchedulingTopology.java [hunk 3]
```diff
@@ -43,11 +43,33 @@
 	 */
 	Optional<SchedulingExecutionVertex> getVertex(ExecutionVertexID executionVertexId);
 
+	/**
+	 * Looks up the {@link SchedulingExecutionVertex} for the given {@link ExecutionVertexID}.
+	 *
+	 * @param executionVertexId identifying the respective scheduling vertex
+	 * @return The respective scheduling vertex or throw {@link Exception} if the vertex does not exist
```
Comment (GJL) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/SchedulingTopology.java]: * It is misleading to write that we potentially throw `Exception` because `IllegalArgumentException` is unchecked. 
* You can use @throws tag

```
* @return The respective scheduling vertex
* @throws IllegalArgumentException If the vertex does not exist
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/SchedulingTopology.java [hunk 4]
```diff
@@ -43,11 +43,33 @@
...
 	 * @param intermediateResultPartitionId identifying the respective scheduling result partition
 	 * @return Optional containing the respective scheduling result partition or none if the partition does not exist
 	 */
 	Optional<SchedulingResultPartition> getResultPartition(IntermediateResultPartitionID intermediateResultPartitionId);
+
+	/**
+	 * Looks up the {@link SchedulingResultPartition} for the given {@link IntermediateResultPartitionID}.
+	 *
+	 * @param intermediateResultPartitionId identifying the respective scheduling result partition
+	 * @return The respective scheduling result partition or throw {@link Exception} if the partition does not exist
```
Comment (GJL) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/SchedulingTopology.java]: See my other comment about `Exception`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/strategy/TestingSchedulingTopology.java (hunk 2 above)
Comment (eaglewatcherwb) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/strategy/TestingSchedulingTopology.java]: OK

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/SchedulingTopology.java (hunk 3 above)
Comment (eaglewatcherwb) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/SchedulingTopology.java]: OK

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/SchedulingTopology.java (hunk 4 above)
Comment (eaglewatcherwb) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/SchedulingTopology.java]: OK

**Comment by GJL** [comment]
Merging.
